File: 1129-shortest-path-with-alternating-colors/1129-shortest-path-with-alternating-colors.java
class Pair{
    int dest;
    int color;
    public Pair(int dest,int color){
        this.dest = dest;
        this.color = color;
    }
}

class Tuple{
    int node,dist,lastColor;

    public Tuple(int node,int lastColor,int dist){
        this.node = node;
        this.lastColor = lastColor;
        this.dist = dist;
    }
}

class Solution {
    public int[] shortestAlternatingPaths(int n, int[][] redEdges, int[][] blueEdges) {
        ArrayList<ArrayList<Pair>> adj = new ArrayList<>();

        for(int i=0;i<n;i++){
            adj.add(new ArrayList<>());
        }

        // 0-> red, 1-> blue
        for(int edge[]:redEdges){
            int u = edge[0];
            int v = edge[1];

            adj.get(u).add(new Pair(v,0));

        }

        for(int edge[]:blueEdges){
            int u = edge[0];
            int v = edge[1];

            adj.get(u).add(new Pair(v,1));
        }

        int vis[][] = new int[n][2];
        //0->red,1->blue
        int dist[] = new int[n];
        Arrays.fill(dist,-1);
        dist[0] = 0;

        Queue<Tuple> q = new LinkedList<>();

        q.add(new Tuple(0,-1,0));

        while(!q.isEmpty()){
            Tuple it = q.poll();
            int node = it.node;
            int lastColor = it.lastColor;
            int distance = it.dist;

            for(Pair edge:adj.get(node)){
                int color = edge.color;
                int adjNode = edge.dest;

                if(color == lastColor){
                    continue;
                }

                if(vis[adjNode][color] == 1){
                    continue;
                }
                vis[adjNode][color] = 1;

                if(dist[adjNode] == -1){
                    dist[adjNode] = distance+1;
                }

                q.add(new Tuple(adjNode,color,distance+1));
            }
        }

        return dist;
    }
}

File: 1129-shortest-path-with-alternating-colors/1129-shortest-path-with-alternating-colors.cpp
class Solution {
public:
    vector<int> shortestAlternatingPaths(int n, vector<vector<int>>& redEdges, vector<vector<int>>& blueEdges) {
        vector<vector<pair<int,int>>> adj(n);

        // 0 -> red
        // 1-> blue

        for(auto edge: redEdges){
            int u = edge[0];
            int v = edge[1];

            adj[u].push_back({v,0});
        }

        for(auto edge: blueEdges){
            int u = edge[0];
            int v = edge[1];

            adj[u].push_back({v,1});
        }

        vector<vector<int>> vis (n,vector<int> (2,0));

        vector<int> dist(n,-1);
        dist[0] = 0;
        queue<pair<int,pair<int,int>>> q;
        q.push({0,{-1,0}});
        // node,lastcolor,dist

        while(!q.empty()){
            int node = q.front().first;
            int lastColor = q.front().second.first;
            int distance = q.front().second.second;

            q.pop();

            for(auto it: adj[node]){
                int adjNode = it.first;
                int color = it.second;

                if(color == lastColor) continue;

                if(vis[adjNode][color] == 1) continue;

                int newDist = distance + 1;
                if(dist[adjNode] == -1){
                    dist[adjNode] = newDist;
                }

                vis[adjNode][color] = 1;
                q.push({adjNode,{color,newDist}});
            }
        }

        return dist;
    }
};